time.Sleep in integration tests — comment and waitRun/waitInspect

vdemeester opened the pull request:
Following the discussion in https://github.com/docker/docker/pull/13501#discussion_r36549317 , remove what seems unnecessary time.Sleep (1 second even) and comment the ones that seemed necessary.
- The comments are also here to see if the `Sleep` time is not too much (for these ones).
- Sometimes `time.Sleep(1 * time.Second)` and sometimes `time.Sleep(time.Second)`, update it to make them all look the same.

🐸

<del>Edit : The lines I wanted to remove are needed to be sure the image are build in a correct order so… I reduced the time to see if it still works — but it makes this PR a little more worthless so feel free to close while I'm sleeping :sweat_smile:.</del>


Signed-off-by: Vincent Demeester vincent@sbr.pm

LK4D4 commented on integration-cli/docker_api_containers_test.go:
```diff
@@ -432,7 +432,7 @@ func (s *DockerSuite) TestGetStoppedContainerStats(c *check.C) {
 
 	// allow some time to send request and let daemon deal with it
-	time.Sleep(1 * time.Second)
+	time.Sleep(time.Second)
```
I don't believe that this change makes something better :) It was easier to change it before.

vdemeester replied:
@LK4D4 yeah it's just to make all *sleep 1 second* looks the same. ``time.Sleep(1 * time.Second)`` or ``time.Sleep(time.Second)`` I don't really have a strong opinion :sweat_smile:. 

Maybe I'm a little OCD on that though :sweat: but as I had all the ``time.Sleep`` in integration-cli in front of me, so it caught my eyes :blush:.

vdemeester replied:
But yeah ``1 * time.Second`` makes it easier to change after so.. I'll make the inverted change :wink:.

duglin commented on integration-cli/docker_cli_links_test.go:
```diff
@@ -138,6 +138,7 @@ func (s *DockerSuite) TestLinksHostsFilesInject(c *check.C) {
 	idTwo := strings.TrimSpace(out)
 
+	// Wait 1 second to make sure both container started
 	time.Sleep(1 * time.Second)
```
Can we replace this sleep with a couple of `waitRun(id))`'s   ?

duglin commented on integration-cli/docker_cli_restart_test.go:
```diff
@@ -31,6 +31,7 @@ func (s *DockerSuite) TestRestartRunningContainer(c *check.C) {
 	cleanedContainerID := strings.TrimSpace(out)
 
+	// Wait 1 second to make sure the container is started
 	time.Sleep(1 * time.Second)
```
waitRun(id) again?

duglin commented on integration-cli/docker_cli_restart_test.go:
```diff
@@ -42,6 +43,7 @@ func (s *DockerSuite) TestRestartRunningContainer(c *check.C) {
 	out, _ = dockerCmd(c, "logs", cleanedContainerID)
 
+	// Wait 1 second to make sure the container has restart
 	time.Sleep(1 * time.Second)
```
waitRun(id)?

duglin commented on integration-cli/docker_cli_run_test.go:
```diff
@@ -2362,6 +2362,7 @@ func (s *DockerSuite) TestRunPidHostWithChildIsKillable(c *check.C) {
 	dockerCmd(c, "run", "-d", "--pid=host", "--name", name, "busybox", "sh", "-c", "sleep 30; echo hi")
 
+	// Wait 1 second to make sure the container is started
 	time.Sleep(1 * time.Second)
```
waitRun(id)?

duglin commented on integration-cli/docker_cli_wait_test.go:
```diff
@@ -20,7 +20,8 @@ func (s *DockerSuite) TestWaitNonBlockedExitZero(c *check.C) {
 		status, err = inspectField(containerID, "State.Running")
```
maybe replace this entire block with a waitInspect() call?

duglin commented on integration-cli/docker_cli_wait_test.go:
```diff
@@ -73,7 +74,8 @@ func (s *DockerSuite) TestWaitNonBlockedExitRandom(c *check.C) {
 		status, err = inspectField(containerID, "State.Running")
```
waitInspect() ?

duglin commented on integration-cli/docker_utils.go:
```diff
@@ -1238,6 +1238,7 @@ func runCommandAndReadContainerFile(filename string, cmd *exec.Cmd) ([]byte, err
 		return nil, fmt.Errorf("%v: %q", err, out)
 	}
 
+	// Wait 1 second to make sure the container is started
```
waitRun()?

duglin commented on integration-cli/docker_cli_attach_unix_test.go:
```diff
@@ -73,6 +73,7 @@ func (s *DockerSuite) TestAttachAfterDetach(c *check.C) {
 		close(errChan)
 	}()
 
+	// Wait 500 millisecond to make sure the container started
```
waitRun() ?

duglin commented:
on the `time.Second` vs `1 * time.Second` thingy... I'm meh  :-)

vdemeester commented on integration-cli/docker_cli_wait_test.go:
```diff
@@ -93,9 +77,6 @@ func (s *DockerSuite) TestWaitBlockedExitRandom(c *check.C) {
 	if err := waitRun(containerID); err != nil {
 		c.Fatal(err)
 	}
-	if err := waitRun(containerID); err != nil {
```
There was two ``waitRun(…)`` following :sweat:.

duglin commented on integration-cli/docker_cli_wait_test.go:
```diff
@@ -67,16 +59,8 @@ func (s *DockerSuite) TestWaitNonBlockedExitRandom(c *check.C) {
-			c.Fatal("Container should have stopped by now")
-		}
+	if err := waitInspect(containerID, "{{.State.Running}}", "false", 1); err != nil {
+		c.Fatal(err)
```
up above on line 18 you showed a more friendly error msg, can we do that that too? When its not as simple as "previous cmd should have worked" I think it'll help others who come later to know what's going on if we show the "Container should have stopped by now" text.

vdemeester replied:
Ohh you're right >.< I'll do that :wink:

duglin commented:
Just one minor request w.r.t a Fatal msg, but aside from that LGTM

vdemeester commented:
@duglin Oh most of the time waitRun is used like : ``c.Assert(waitRun(id), check.IsNil)`` (like [here](https://github.com/docker/docker/blob/master/integration-cli/docker_api_containers_test.go#L1278)). I could update the rest of them too in this PR maybe, WDYT ?

duglin commented:
I don't think that's as important as the other one because the phrase
"waitRun" itself is pretty clear.  The other case jumped out at me because
it may not be clear to someone new looking at it that we're actually
waiting for the container to stop. But I won't object to you changing these
if you want.

On Sun, Aug 9, 2015 at 3:53 PM, Vincent Demeester <notifications@github.com>
wrote:

> @duglin <https://github.com/duglin> Oh most of the time waitRun is used
> like : c.Assert(waitRun(id), check.IsNil) (like here
> <https://github.com/docker/docker/blob/master/integration-cli/docker_api_containers_test.go#L1278>).
> I could update the rest of them too in this PR maybe, WDYT ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/15412#issuecomment-129232355>.
>

duglin commented:
still LGTM

cpuguy83 commented on integration-cli/docker_cli_images_test.go:
```diff
@@ -25,14 +25,14 @@ func (s *DockerSuite) TestImagesOrderedByCreationDate(c *check.C) {
 		c.Fatal(err)
 	}
-	time.Sleep(time.Second)
+	time.Sleep(1 * time.Second)
```
Do we know why these sleeps are here?

vdemeester replied:
@cpuguy83 yes. This tests test the order of the ``images`` command. It build 3 images in order and wait 1 seconds between each to be sure the datetime difference is big enough.

I should add a comment on thoses (I though I did but.. seems like I did not… :sweat:).

cpuguy83 commented on integration-cli/docker_cli_events_test.go:
```diff
@@ -20,12 +20,12 @@ func (s *DockerSuite) TestEventsTimestampFormats(c *check.C) {
 
 	// Start stopwatch, generate an event
-	time.Sleep(time.Second) // so that we don't grab events from previous test occurred in the same second
+	time.Sleep(1 * time.Second) // so that we don't grab events from previous test occured in the same second
```
There are two sleep here, the 2nd one seems redundant

cpuguy83 commented on integration-cli/docker_cli_events_test.go:
```diff
@@ -223,7 +223,7 @@ func (s *DockerSuite) TestEventsImageUntagDelete(c *check.C) {
 
 func (s *DockerSuite) TestEventsImageTag(c *check.C) {
-	time.Sleep(time.Second * 2) // because API has seconds granularity
+	time.Sleep(2 * time.Second) // because API has seconds granularity
```
This should be only 1 second.

cpuguy83 commented:
The events tests seems like they need their own suite so we can add a `time.Sleep(1 * time.Second)` in the test setup.... assuming that the comments on the sleeps are accurate.

vdemeester committed:
Remove/Comment time.Sleep in integration tests

Remove what seems unnecessary time.Sleep (1 second even) and comment the
ones that seemed necessary.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

vdemeester commented:
@cpuguy83 The strange thing is, there is not that much *event* tests that are doing the ``time.Sleep(…)`` though..

vdemeester commented:
@cpuguy83 so I tried it out and the `time.Sleep` are need on those events test (at least the way the test are written). If we setup a `DockerEventSuite` with a `time.Sleep`` this would mean we wait more (because there is not that much event integration tests that are *sleeping* at start — so I don't know what to do :sweat_smile: :blush: .

cpuguy83 commented on integration-cli/docker_cli_events_test.go:
```diff
@@ -20,12 +20,12 @@ func (s *DockerSuite) TestEventsTimestampFormats(c *check.C) {
+	time.Sleep(1 * time.Second) // so that we don't grab events from previous test occured in the same second
 	start := daemonTime(c)
-	time.Sleep(time.Second) // remote API precision is only a second, wait a while before creating an event
+	// time.Sleep(1 * time.Second) // remote API precision is only a second, wait a while before creating an event
```
Need to remove the commented one :)

vdemeester replied:
ah right :sweat_smile:.

cpuguy83 commented:
LGTM
